src/blockarrayinterface.jl

```diff
@@ -15,6 +15,7 @@ function axes(D::Diagonal{<:Any,<:AbstractBlockVector})
     a = axes(parent(D),1)
     (a,a)
 end
+ArrayLayouts.diagonallayout(lay::AbstractBlockLayout) = DiagonalLayout{typeof(lay)}()
```

Comment:
Because this relies on `AbstractBlockLayout`, I think you have to change the load order here: https://github.com/JuliaArrays/BlockArrays.jl/blob/d270f95e217f8c3bf6249cc4e3cebeb7cd29294f/src/BlockArrays.jl#L51 to do `include("blocklinalg.jl")` before `include("blockarrayinterface.jl")`, as you get an undefined error while compiling otherwise.